Implements the TaskManager API

tcsc opened the pull request:
Adds an implementation & tests for the TaskManager API. In order to do
this,  this patch
 * Adds the notion of a Task Registry; an in-memory task database
   protected by a read-write lock that the service uses to manage
   tasks, and allocate IDs.
 * Wraps the Task Registry in a TaskManagerServer implementation
   defined in the `taskmanager` package, which implements the service
   calls nterms of operations on the task registry and task instances
 * Moves the cmd_test package out from under task and into the top level
   of the project, to signal that they are use outside the `task`
   package tests
 * Adds reasnably comprehensive tests of the taskmanager
   implementation

Note that this patch does not contain an application for running the
server and exposing it over the network. This is coming in a subsequent
patch.

tcsc committed:
Implements the TaskManager API

Adds an implementation & tests for the TaskManager API. In order to do
this,  this patch
 * Adds the notion of a Task Registry; an in-memory task database
   protected by a read-write lock that the service uses to manage
   tasks, and allocate IDs.
 * Wraps the Task Registry in a TaskManagerServer implementation
   defined in the `taskmanager` package, which implements the service
   calls nterms of operations on the task registry and task instances
 * Moves the cmd_test package out from under task and into the top level
   of the project, to signal that they are use outside the `task`
   package tests
 * Adds reasnably comprehensive tests of the taskmanager
   implementation

Note that this patch does not contain an application for running the
server and exposing it over the network. This is coming in a subsequent
patch.

r0mant commented on registry/registry.go:
```diff
@@ -0,0 +1,93 @@
+
+// Registry models a simple task registration system. For a simple exercise
+// like this we are just keeping the task information in memory. This would
+// obviously require somthing more durable in a production server.
```
something

tcsc replied:
Fixed

r0mant commented on registry/registry.go:
```diff
@@ -0,0 +1,93 @@
+}
+
+// Register attempts to register a task with the registry. Keeps trying until
+// it succeeds in generateing a unique ID, or the supplied context expires.
```
generating

tcsc replied:
Refactored out of existence

r0mant commented on registry/registry.go:
```diff
@@ -0,0 +1,93 @@
+	// we have a unique ID registered in the database. To handle the even more
+	// remote possibility of repeated collisions, we will only keep trying
+	// until the supplied context expires
+	for ctx.Err() == nil {
```
What's this check for? If you mean to check for a completed context (to detect an exit for example), normally a select is used:

```go
select {
case <-ctx.Done():
  // handle exit
}
```

tcsc replied:
My basic idea was that the function would keep trying to generate a unique ID until the context completed, rather than just using an arbitrary number of retries. I wasn't sure how to detect that a context was closed, and polling the error  seemed like a good idea at the time. 

I have replaced with a `select` over `ctx.Done()` that has a `default` case, so we don't actually block on the channel.

tcsc replied:
Refactored away, in any case

r0mant commented on registry/registry.go:
```diff
@@ -0,0 +1,93 @@
+			return candidate, nil
+		}
+	}
+	return nullTaskHandle, ctx.Err()
```
Why is this returning the context error?

tcsc replied:
I'm not sure what else to return, other than a generic "registration failed" error. My reasoning is that the only way we could get to this line is that the context has been completed before we have generated an ID, so returning the Cancelled|Timed Out error value from the context seems like a useful bit of information for the caller to get.

I don't actually know if there are any good idioms for this sort of thing; if there are please point them my way :-)

r0mant commented on registry/registry.go:
```diff
@@ -0,0 +1,93 @@
+	// the task ID. This should be statistically unique - certainly we
+	// should not expect to see a collision in the lifetime of the server.
+	//
+	// To handle the remote possibility of a collision, we keep trying until
```
I honestly think this is an overkill but the retry in this case is simple enough.

russjones replied:
I think you can remove and simplify this substantially. A good implementation of UUIDv4 (which you are using) should be polling your OS CSPRNG. If it's giving you potentially duplicate values for a 122-bit number, you're going to see substantially larger issues than this.

This will also have the benefit of making `StartTask` in the `taskmanager` package simpler

tcsc replied:
Yep. I was concentrating on the IDs being arbitrary strings, and stringifying a UUID was just a way of generating a reasonably long, reasonably random string. But in retrospect it's silly to not just trust that they are unique.

r0mant commented on task/task.go:
```diff
@@ -66,6 +67,15 @@ func New(binary string, workingDir string, env map[string]string, args ...string
 	return &t
 }
 
+// String implements Stringer for Task, which is required for not tripping the
```
Where is the race detected exactly? TBH locking in the String() method feels like a hack. Any way to fix the Task implementation instead to not trigger the race detector?

tcsc replied:
It was in the mocks for the `TaskManager`. Whenever the mocking framework tried to format am error message using a Task instance it tripped the race detector; I'm assuming because the default implementation was non-atomically reading several fields in order to generate the string.

The tests that needed this via the mocking framework have now been deleted, so this has gone as well.

r0mant commented on taskmanager/server.go:
```diff
@@ -0,0 +1,155 @@
+
+// Error is an error code describing things that can go wrong with the
+// TaskManager API service.
+type Error int
```
If you want to define a custom error, in Go you'd normally use a struct, like:

```go
type NoSuchTaskError struct {
  Message string
}

func (e *NoSuchTaskError) Error() string {
  retur e.Message
}
```

tcsc replied:
Replaced integer code with something similar to the above.

r0mant commented on taskmanager/server.go:
```diff
@@ -0,0 +1,155 @@
+}
+
+// TaskRegistry abstracts out the desinition of the task0id-to-task mapping, so
+// we can mock it ut for testing
```
<strike>ut</strike>

tcsc replied:
Refactored away

r0mant commented on taskmanager/server.go:
```diff
@@ -0,0 +1,155 @@
+		req.GetEnvironment(),
+		req.GetArgs()...)
+
+	// Start the task *before* registering it, as we don't want to expose the task
```
I think this may expose the client to unexpected side-effects, for example what if the task was started but failed to register? It will basically be left orphaned with the client having no ability to know this and/or stop it. Is there a way to handle this scenario better?

tcsc replied:
The task gets killed if registration fails a few lines down.

The other alternative I thought of was essentially registering a `nil` task to allocate a unique ID. Once we have that, simply assigning anew task to the empty slot cannot fail, so the flow would look something like this:

1. Get a unique task ID, which could potentially fail
2. On success, create the task, which might fail
3. On success, bind the task to the Id, which cannot fail
4. Fail all RPC calls that return a `nil` task with `no-such-task`

This trades off simplicity here against having the occasional bit of `nil` garbage in the task registry; which is probably the right trade given the exercise.

That feels a bit better to me.

r0mant commented on taskmanager/server.go:
```diff
@@ -0,0 +1,155 @@
+func (server *Server) FetchLogs(
+	ctx context.Context, req *api.FetchLogsRequest) (*api.FetchLogsResponse, error) {
+
+	task := server.registry.Lookup(req.TaskId.Id)
```
More of a nit, but looking at this I think the contract would be a little bit cleaner if the registry's Lookup method just returned a not found error itself rather than translating "nil" to "not found" in server.

tcsc replied:
I'm on the fence with this one. I can see what you're saying, but I also like the separation of concerns, where the registry is an internal component that does what it does, while the `Server` is responsible for deciding exactly what errors get propagated to the client.

 It feels a bit like leaking implementation details.

I'm going to leave it as-is for now, but happy for further guidance.

russjones replied:
This is a minor issue, no need to make changes on this, but since you asked, take a look at the following article.

https://en.wikipedia.org/wiki/Semipredicate_problem

I know this is one of the things I disliked about C myself.

r0mant commented on taskmanager/server.go:
```diff
@@ -0,0 +1,155 @@
+		return NoSuchTask
+	}
+
+	// Note the hardcoded 5s timeout here; this shoud at the very least be
```
should

tcsc replied:
Fixed

r0mant commented on taskmanager/server.go:
```diff
@@ -0,0 +1,155 @@
+	// Note the hardcoded 5s timeout here; this shoud at the very least be
+	// a be parameter of the Server, preferably configurable somehow by the
+	// user. For the sake of this exercise, it's just a hardcoded value.
+	signalCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```
Any reason the context passed to this server method isn't used as a parent here?

tcsc replied:
Because the signal timeout (which is implemented as waiting on a `context.Done()`) is not bounded by the lifetime of this function call.

Once this call sends the signal its done, and will return to the caller. The caller may then (and in fact, probably _will_) cancel the enclosing context. If the signal context was a child of the enclosing context, this would end up killing the task early.

r0mant commented on taskmanager/server.go:
```diff
@@ -0,0 +1,155 @@
+	// RPC calls
+	id, err := server.registry.Register(ctx, t)
+	if err != nil {
+		killTask(t) // Don't leak the task if we fail to register it
```
Pass context to this method?

tcsc replied:
Refactored away

tcsc committed:
Fixes infinite loop in test.

The `trap` directive in `sh` apparently works differently in linux vs
OSX, where the `SIG` prefix is allowed. This patch removes the prefix to
make the test shell script work under linux.

Also adds a timeout to the test case to prevent the infinite loop if the
test case ever breaks again.

tcsc committed:
Refactors Task Registry interface

tcsc committed:
Dead code removal

fspmarshall commented on registry/registry.go:
```diff
@@ -0,0 +1,125 @@
+var UnallocatedIDError = errors.New("ID not allocated")
+var AlreadyBoundIDError = errors.New("ID already bound to a task")
+
+// BindToID binds a task to a pre-allocated ID. It is considered an error to
+// attempt to bind a tsk to an ID that has not been pre-allocated, or to an
+// ID already bound to a task.
+func (registry *Registry) BindToID(id string, t *task.Task) error {
```
Whats the usecase for having IDs be separately allocated in advance?  Seems like this one feature adds a lot of complexity.

tcsc replied:
It does. Most of it stems from me treating the task IDs as an arbitrary string, rather than a stringified UUID.

If you think of the ID as an arbitrary string, then the ID generation could potentially fail (e.g. if you had _really_ bad luck and continually generated clashing ID strings until the context timed out). Anyway, I didn't want to have to start the task process and then immediately kill it if the registration failed, so I split the registration into two parts; allocating the ID in advance (which could fail) and binding the task to the ID once it was started (which could not fail, at least if used correctly).

But yeah, it's less than ideal.

Following @russjones' suggestion to just trust that the generated UUIDs are actually unique drastically simplifies the whole thing and pretty much all of this is now deleted.

tcsc committed:
Trust that UUIDs are unique

tcsc committed:
typo fix

russjones commented on taskmanager/server.go:
```diff
@@ -0,0 +1,125 @@
+
+	var exitCode *int32
+	status, taskExitCode := task.Status()
+	if taskExitCode != -1 {
```
Why not use a constant here?
